projects/l2_pooling/noise_tolerance_l2_l4.py

```diff
@@ -0,0 +1,357 @@
+  See 'createRandomObjects'.
+
+  @param noiseLevels (list of floats)
+  The noise levels to experiment with. The experiment is run once per noise
+  level. Noise is applied at a constant rate to exactly one cortical column.
+  It's applied to the same cortical column every time, and this is the cortical
+  column that is measured.
+
+  @param featureDropout (bool)
+  Whether to use a noisy feature
+
+  @param locationDropout (bool)
+  Whether to use a noisy location
+  """
+
+  featureSDR = lambda : set(random.sample(xrange(2048), 40))
+  locationSDR = lambda : set(random.sample(xrange(1024), 40))
+
+  featureSDRsByColumn = [defaultdict(featureSDR) for _ in xrange(numColumns)]
+  locationSDRsByColumn = [defaultdict(locationSDR) for _ in xrange(numColumns)]
+
+  exp = L4L2Experiment(
+    "Experiment",
+    numCorticalColumns=numColumns,
+    externalInputSize=1024,
+    seed=random.randint(2048, 4096)
+  )
+
+  exp.learnObjects(
+    dict((objectName,
+          [dict((column,
+                 (locationSDRsByColumn[column][location],
+                  featureSDRsByColumn[column][features[location]]))
+                for column in xrange(numColumns))
+           for location in xrange(len(features))])
+         for objectName, features in objects.iteritems()))
+
+  if locationDropout:
+    # Now that the objects are learned, allow bursting L4 minicolumns to
+    # activate cells in L2.
+    exp.L4Columns[0].setParameter("defaultOutputType", 0, "active")
+
+  results = defaultdict(list)
+
+  for noiseLevel in noiseLevels:
+    # Try to infer the objects
+    for objectName, features in objects.iteritems():
+      exp.sendReset()
+
+      inferredL2 = exp.objectL2Representations[objectName]
+
+      sensorPositionsIterator = greedySensorPositions(numColumns, len(features))
+
+      # Touch each location at least numInitialTouches times, and then touch it
+      # once more, testing it. For each traversal, touch each point on the object
+      # ~once. Not once per sensor -- just once. So we translate the "number of
+      # traversals" into a "number of touches" according to the number of sensors.
+      numTouchesPerTraversal = len(features) / float(numColumns)
+      numInitialTouches = int(math.ceil(numInitialTraversals * numTouchesPerTraversal))
+      numTestTouches = len(features)
+
+      for touch in xrange(numInitialTouches + numTestTouches):
+        sensorPositions = next(sensorPositionsIterator)
+
+        sensation = dict(
+          (column,
+           (locationSDRsByColumn[column][sensorPositions[column]],
+            featureSDRsByColumn[column][features[sensorPositions[column]]]))
+          for column in xrange(1, numColumns))
+
+        for _ in xrange(TIMESTEPS_PER_SENSATION):
+          # Add noise to the first column.
+          featureSDR = featureSDRsByColumn[0][features[sensorPositions[0]]]
+          if featureDropout:
+            featureSDR = withDropout(featureSDR, noiseLevel)
```

Comment:
I guess the newly introduced bits won't affect the result much. But it is still good to have them, otherwise it is more like "tuning down signal strength" rather than "add noise".